decidim-proposals/app/permissions/decidim/proposals/permissions.rb

```diff
@@ -0,0 +1,143 @@
+# frozen_string_literal: true
+
+module Decidim
+  module Proposals
+    class Permissions
+      def self.allows?(user, action, context)
+        new(user, action, context).allowed?
+      end
+
+      def initialize(user, action, context)
+        @user = user
+        @action = action
+        @context = context
+      end
+
+      def allowed?
+        # Stop checks if the user is not authorized to perform the
+        # action for this space
+        return false unless spaces_allows_user?
+        return false unless user
+
+        # Delegate the admin permission checks to the admin permissions class
+        return Decidim::Proposals::Admin::Permissions.allows?(user, action, context) if action[:scope] == "admin"
+        return false if action[:scope] != "public"
+
+        return false if action[:subject] != "proposal"
+
+        return true if can_create_proposal?
+        return true if can_edit_proposal?
+        return true if can_withdraw_proposal?
+        return true if action[:action] == "report"
+
+        return true if can_endorse_proposal?
+        return true if can_unendorse_proposal?
+
+        return true if can_vote_proposal?
+        return true if can_unvote_proposal?
+
+        false
+      end
+
+      private
+
+      attr_reader :user, :action, :context
+
+      def spaces_allows_user?
+        space.manifest.permissions_class&.allows?(user, action, context)
+      end
+
+      def current_settings
+        @current_settings ||= context.fetch(:current_settings, nil)
+      end
+
+      def component_settings
+        @component_settings ||= context.fetch(:component_settings, nil)
+      end
+
+      def component
+        @component ||= context.fetch(:current_component, nil)
+      end
+
+      def space
+        @space ||= component&.participatory_space
```

Comment:
Is this safe navigation actually safe? :S. All components have a parent space.

Reply:
Well, it *is* safe. It's just taking care of the improbable case where a component does not have a participatory space, I can remove this safe navigation.

Reply:
Also, the `component` method can return `nil`

Reply:
How can `component` be nil in the context of proposals though? Shouldn't we enforce this?

Reply:
I've removed this safe navigation!